### DISCUSSION TITLE
Support of replacing assert with something else

### DISCUSSION THREAD

**Discussion by Anton-V-K** [nlohmann/json#2239] [Ideas]
The library uses `assert` in certain places to signal about fatal errors.
An example:
```
    void assert_invariant() const noexcept
    {
        assert(m_type != value_t::object or m_value.object != nullptr);
        assert(m_type != value_t::array or m_value.array != nullptr);
        assert(m_type != value_t::string or m_value.string != nullptr);
        assert(m_type != value_t::binary or m_value.binary != nullptr);
    }
```
Such kind of signal isn't friendly (when debugging, the code just crashes) if compared to `_ASSERT` (from `<crtdbg.h>`, which is available for Microsoft C++ compiler) or similar macros (like https://github.com/foonathan/debug_assert).
Please, provide a way to use another macro instead of `assert`.

It is not possible to globally redefine `assert` for the library, because it includes `<cassert>` directly.

#### Which compiler and operating system are you using?

- Compiler: Microsoft C++ compiler
- Operating system: Windows 7 64-bit

#### Which version of the library did you use?

- [x] latest release version 3.7.3
- [ ] other release - please state the version: ___
- [ ] the `develop` branch

**Comment by nlohmann** [comment]
Thanks for reporting. I think using a macro like `JSON_ASSERT` than can be redefined by client code should be possible. I will have a look.

**Reply by nlohmann** [reply]
See https://github.com/nlohmann/json/pull/2242. Feedback welcome!

**Reply by nlohmann** [reply]
The feature has now been merged to the `develop` branch and will be part of the next release.
